compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -222,10 +224,20 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     bug!("Found {operand_kind:?} for operand {operand:?}");
                 };
                 if let OperandValueKind::Immediate(out_scalar) = cast_kind {
-                    let cast_bty = bx.backend_type(cast);
-                    Some(OperandValue::Immediate(Self::transmute_immediate(
-                        bx, imm, in_scalar, out_scalar, cast_bty,
-                    )))
+                    match (in_scalar, out_scalar) {
+                        (ScalarOrZst::Zst, ScalarOrZst::Zst) => {
+                            Some(OperandRef::new_zst(bx, cast).val)
+                        }
+                        (ScalarOrZst::Scalar(in_scalar), ScalarOrZst::Scalar(out_scalar))
+                            if in_scalar.size(self.cx) == out_scalar.size(self.cx) =>
```

Comment:
Why is this check necessary?

For pairs you need to check the inner size because matching size of the whole pair doesn't necessarily imply matching sizes of elements... But surely the `operand.layout.size != cast.size` check is sufficient for scalars? 🤔

Reply:
Oh, good call.  Let me give that a shot -- might allow deleting more code too...

Reply:
Removed this check, but `ScalarOrZst::size` is still needed elsewhere, so I couldn't remove more than that.

Reply:
Oh, no, vectors.  Without this check `<4 x i64>` to `<8 x f32>` trips the later debug-assert.

I suppose I could hypothetically `bitcast` that anyway, but I'm not doing that this PR.

Added some codegen tests for vectors so that when future me breaks this while running `--keep-stage-std` again, I'll notice without CI 😬